### PR TITLE
Align to GeoNode master (4.2.0) and improve/fix docker initialization

### PR DESCRIPTION
**Pull request by afabiani**
This PR introduces improvements and fixes to docker-compose and initialization scripts in general:

- Removing some unused variables.
- Removing some unused and obsolete containers, such as Jenkins.
- Removing the incorrectly written and non-functioning part for setting the password of GeoServer.
- Inverting the startup dependency from GeoServer -> Django to Django -> GeoServer.
- Ensuring that OAUTH2 is properly initialized on restart in all its parts.
- Enhancing NGINX to make sure that GeoServer functions well behind any proxy, including HTTPS.

Some implementation details:

1. Environment file

`DOCKERHOST` and `DOCKER_HOST_IP` variables removed as useless and misleading. Instead, it exposed `NGINX_BASE_URL` which was calculated automatically and badly.

The variables
```
GEONODE_LB_HOST_IP=django
GEONODE_LB_PORT=8000
GEOSERVER_LB_HOST_IP=geoserver
GEOSERVER_LB_PORT=8080
```
are important, they are used by `NGINX` to create the upstream `proxyPass` and by GeoServer to create the internal `OAuth2` endpoints. They represent the name and port of the GeoNode and GeoServer containers. In general it is not necessary to change them, but in an environment like `Microsoft Azure` these must be set to the address of the `subnet` as the names are not internally resolved.

Increased `MAX_DOCUMENT_SIZE=200` because the previous default of `2Mb` did not allow rendering even one document.

Updated `create-envfile.py` to also enable `staging` mode to test `HTTPS` on `localhost`, which was previously totally ignored.

2. Dockerfiles

Removed useless stuff and started it from `geonode/geonode-base:latest-ubuntu-22.10`

3. Readme

Added a `Quick Start` section with `docker`

4. Docker Compose

Fixed healthchecks which are now faster and more efficient and switched Django -> GeoServer startup dependency, because it's the Django entrypoint that sets `OAutg2` parameters in the `GEOSERVER_DATA_DIR` before it starts.

Updated to `POSTgresql 15`

5. UWSGI + NGINX

Enabled `proxyPass` instead of `uwsgiPass` because we set `headers` correctly in `geonode.conf` only with `proxyPass`

6. GeoServer docker entry points

Fixed the `.sh` to set `OAuth2` which previously worked badly and only partially checking when there was the `localhost` value. Now they update everything and always.

7. nginx/geonode.conf.envsubst

Made `proxy passes` dynamic

`set $upstream $GEOSERVER_LB_HOST_IP:$GEOSERVER_LB_PORT;`

`set $upstream $GEONODE_LB_HOST_IP:$GEONODE_LB_PORT;`

Updated the GeoServer conf so that the GUI works fine behind any proxy.

8. A general reformat to the `black` code

9. Version bump

from `4.0.0.final` to `4.2.0.dev`

### CODE REVIEW

**Comment by ridoo** [comment]
@afabiani these are quite good changes! However, the PR raises a maintenance question:

What do you actually recommend when users have to decide between a GeoNode Fork or the creation of a separate  Geonode project via geonode-project template?

The main problem I see is to keep track of all the changes you make on this repository (e.g. those within this PR). Once a project has been created, it gets detached from version control. If a user decides to upgrade their GeoNode project, they would have to follow the commit history and apply all changes manually, right? Wouldn't it be better to just use a GeoNode fork so the changes become visible once you do a merge/pull?

Is there is a discussion (issue/documentation) about Forking vs. GeoNode-Project somewhere? Some good practice guidance would be helpful too.

I'd be happy for any pointer :-)

**Comment by afabiani** [comment]
Those changes won't ba backported. Those will be eventually part of the 4.2.0 train, currently under development.

**Comment by ridoo** [comment]
Ok .. but what about mainaining a project which got generated from the geonode-project repository? When upgrading to 4.2.0 .. I would have to go through the commit history and apply the changes manually on that project. Or do I miss something?

**Comment by afabiani** [comment]
The idea would be to update the documentation of the 4.2.0 release with a specific section for this topic. However we can open a discussion on the mailing list to further investigate the problem and hear from the community if there are some suitable proposals/ideas.

**Comment by ridoo** [comment]
> However we can open a discussion on the mailing list to further investigate the problem and hear from the community if there are some suitable proposals/ideas.

You mean `geonode-devel@lists.osgeo.org`? I am subscribed, but it seems it makes not much noise in my inbox. However, I will open a discussion thread there.

_edit: mail address .. geonode-devel is the right one -- in fact I only found the geonode-devel-requests address in my inbox_

**Comment by ridoo** [comment]
> However, I will open a discussion thread there.

@afabiani my email awaits approval :man_shrugging:

File: docker/geoserver/tasks.py
Comment (ridoo) [inline docker/geoserver/tasks.py:45]: a strayed localhost, I suspect?!

Comment (afabiani) [reply docker/geoserver/tasks.py:45]: Nope it's ok. It will always refer to the local container.

**Comment by giohappy** [comment]
@ridoo if you agree I would merge the PR

**Comment by ridoo** [comment]
@giohappy if it is ok for you, I'd take a deeper look later this day, before just saying "ok".

**Review by giohappy — APPROVED** [review]

File: .env.sample
Comment (ridoo) [inline .env.sample]: Not sure this review is the right place to discuss this, but what do you think about renaming such variables to the get scope more clear:

Using `LB_GEOSERVER_HOST` for example would indicate scope load balancer (that is what LB stands for, right?). So I would read "Loadbalancer's Geoserver Host". The `_IP` suffix would also be unnecessary IMO.

File: Dockerfile
Comment (ridoo) [inline Dockerfile]: Put this into a single line using `\` would decrease size of Docker image (less layers)

File: Dockerfile
Comment (ridoo) [inline Dockerfile]: Same as above: Single line -> less sized Docker image

File: Dockerfile
Comment (ridoo) [inline Dockerfile]: here again

File: docker-purge.sh
Comment (ridoo) [inline docker-purge.sh:2]: This file is quite dangerous as it removes everything non-geonode related, too! To my experience first-time users do not understand the implications immediately. Having a cleaned Docker setup is afterwards, well, .. surprising, at least.

Leverage the `--filter` flag: 

```
# delete all containers with name including "thuenen" 
docker rm $(docker ps -aq --filter=name=thuenen)
# delete all volumes with name including "thuenen"
docker volume rm $(docker volume ls -q --filter name=thuenen)
# delete all images having a `4.0` tag starting with `geonode`
docker rmi $(docker images -q --filter=reference='geonode/*:4.0')
```

File: docker/geoserver/Dockerfile
Comment (ridoo) [inline docker/geoserver/Dockerfile]: Single line RUN reduces Docker image size

File: docker/geoserver/docker-compose.yml
Comment (ridoo) [inline docker/geoserver/docker-compose.yml:28]: Why not using `SITEURL` here as well? `NGINX_BASE_URL` is actually the same: 

https://github.com/GeoNode/geonode-project/blob/95c0c41fff626ed9bcbbe624e63cda741c315991/.env.sample#L65

In which cases `SITEURL` and `NGINX_BASE_URL` will differ?

File: docker/geoserver/entrypoint.sh
Comment (ridoo) [inline docker/geoserver/entrypoint.sh]: is there a reason to hide such logs from stdout by default `:-FALSE}`?

File: docker/geoserver/entrypoint.sh
Comment (ridoo) [inline docker/geoserver/entrypoint.sh]: To be honest, I am a bit confused about `NGINX_BASE_URL`. It pointed to the nginx container beforehand, but now points to geonode itself. I do not get the intent of it. 

At least the echo statement above should be updated.

**Review by giohappy — APPROVED** [review]

File: docker/geoserver/tasks.py
Comment (ridoo) [reply docker/geoserver/tasks.py]: Remove redundant import

File: docker/nginx/Dockerfile
Comment (ridoo) [inline docker/nginx/Dockerfile]: Run in a single line

File: src/uwsgi.ini
Comment (ridoo) [inline src/uwsgi.ini:2]: Is this the new default? No uwsgi-sockets anymore?

**Review by ridoo — COMMENTED** [review]
I finished review .. however, I touched some parts which relate to Geonode core as well (the PR puts the changes made upstream here) .

However, I did not want to block the PR.

File: .env.sample
Comment (giohappy) [reply .env.sample]: I agree @ridoo I was also considering this renaming, but I was considering this for a new PR. This is already quite big!

File: Dockerfile
Comment (giohappy) [reply Dockerfile:33]: I haven't investigated this thoroughly but I suspect that the second `pip install -e .` is meant to configure the python path for the geonode project's code

File: Dockerfile
Comment (giohappy) [reply Dockerfile]: you mean this?
```
RUN apt-get autoremove --purge &&\ 
    apt-get clean
```

File: docker/geoserver/docker-compose.yml
Comment (giohappy) [reply docker/geoserver/docker-compose.yml:28]: Because `NGINX_BASE_URL` this is what the Nginx and Geoserver images expect. I know we can change them but, again, I would do it in another PR.

File: docker/geoserver/entrypoint.sh
Comment (giohappy) [reply docker/geoserver/entrypoint.sh]: I agree with you, I also think it should TRUE by default

File: src/uwsgi.ini
Comment (giohappy) [reply src/uwsgi.ini:2]: This is a controverial decision. The Nginx configuration, at the moment, sets many `proxy_*` parameters since `http-socket` was used in the past. Then it was migrated to `uwsgi-socket` since it's better if we have a uWsgi upstream. However the `proxy_*` parameteres weren't converted, so in this moment they don't have any effect.

We should convert them to the corresponding `[uwsgi_*](http://nginx.org/en/docs/http/ngx_http_uwsgi_module.html)` version, but it requires testing.

I think we should do it, but I would suggest in another dedicated PR. What do you think?

File: docker/geoserver/entrypoint.sh
Comment (giohappy) [reply docker/geoserver/entrypoint.sh]: You're right. Neither would work since this variable [is used inside Geoserver configuration](https://github.com/GeoNode/geonode-project/blob/align_with_geonode_master/docker/geoserver/set_geoserver_auth.sh#L62-L64) for the public URL of the LB (`Nginx`).

@afabiani We can revert this and sei it to "http://geonode:80" as it was before, although it's wrong.

Comment (afabiani) [reply docker/geoserver/entrypoint.sh]: @giohappy yes, sure.

Comment (giohappy) [reply docker/geoserver/entrypoint.sh]: I propose to set it to [GEOSERVER_LOCATION](https://github.com/GeoNode/geonode/blob/align_with_docker_master/.env.sample#L91) by default:

```bash
if [ -z `echo ${NGINX_BASE_URL} | sed 's/http:\/\/\([^:]*\).*/\1/'` ]
then
    echo "NGINX_BASE_URL is empty so I'll use the static nginx hostname \n"
    echo export NGINX_BASE_URL=http://${GEOSERVER_LOCATION} >> /root/.override_env
    echo "The calculated value is now NGINX_BASE_URL='${GEOSERVER_LOCATION}' \n"
else
    echo "NGINX_BASE_URL is filled so I'll leave the found value '$NGINX_BASE_URL' \n"
fi
```

and we should probably do the reverse: set `GEOSERVER_LOCATION` to [$GEOSERVER_LOCATION](url) if `$NGINX_BASE_URL` is not set. It means: Geoserver location is the default one unless there's an LB in front of it.

File: .env.sample
Comment (ridoo) [reply .env.sample]: just fine with that

File: docker/geoserver/entrypoint.sh
Comment (ridoo) [reply docker/geoserver/entrypoint.sh]: I see .. but what is the difference between `SITEURL` and `NGINX_BASE_URL`? Aren't both pointing to the same external URL? What infrastructure configuration do you have in mind so that the two variables would be different?

File: src/uwsgi.ini
Comment (ridoo) [reply src/uwsgi.ini:2]: Definitely a separate PR

File: docker/geoserver/entrypoint.sh
Comment (afabiani) [reply docker/geoserver/entrypoint.sh]: @ridoo in environments like Azure the docker containers names are not resolved and therefore we need to provide here the ip address or some other internal reference.

Comment (ridoo) [reply docker/geoserver/entrypoint.sh]: @afabiani I do not know Azure very much, so just fine with that. But which one is an internal config? As I understood @giohappy correctly, both point to the same value having just a different name in different containers.

Comment (afabiani) [reply docker/geoserver/entrypoint.sh]: @ridoo genrally speaking this variable should **always** point to the `SITEURL` value, at least on a production like environment; if pointing to the containers (assuming that you started the NGINX one too) it must point to `geonode:80` otherwise it should point to whatever is currently proxying the `http(s)` requests.

Comment (afabiani) [reply docker/geoserver/entrypoint.sh]: > I propose to set it to [GEOSERVER_LOCATION](https://github.com/GeoNode/geonode/blob/align_with_docker_master/.env.sample#L91) by default:
> 
> ```shell
> if [ -z `echo ${NGINX_BASE_URL} | sed 's/http:\/\/\([^:]*\).*/\1/'` ]
> then
>     echo "NGINX_BASE_URL is empty so I'll use the static nginx hostname \n"
>     echo export NGINX_BASE_URL=http://${GEOSERVER_LOCATION} >> /root/.override_env
>     echo "The calculated value is now NGINX_BASE_URL='${GEOSERVER_LOCATION}' \n"
> else
>     echo "NGINX_BASE_URL is filled so I'll leave the found value '$NGINX_BASE_URL' \n"
> fi
> ```
> 
> and we should probably do the reverse: set `GEOSERVER_LOCATION` to [$GEOSERVER_LOCATION](url) if `$NGINX_BASE_URL` is not set. It means: Geoserver location is the default one unless there's an LB in front of it.

@giohappy the `NGINX_BASE_URL` should be the `GEONODE_PUBLIC_LOCATION` not the `GEOSERVER_LOCATION`

Comment (giohappy) [reply docker/geoserver/entrypoint.sh]: I've done this way @afabiani and @ridoo https://github.com/GeoNode/geonode-project/commit/3393da2c569353d62e5ffcc478853a46c5b3c297.

However, the naming of "LB*' variable is still confusing. The name refers to a Load Balancer, but at the moment we're using them for the Docker network hostnames, while the LB URLs are `SITEURL`, `NGINX_BASE_URL` and the root fo `GEOSERVER_PUBLIC_LOCATION`.

I know the reason is the history of these files and configurations. In a future PR I would propose to review them further by normalization, deduplication, and renaming.

Comment (ridoo) [reply docker/geoserver/entrypoint.sh]: Sounds like a good idea. I scratched my head several times. But as you said: sometimes such things on their own slightly diverge from their original intent.

**Pull request merged**
